pydantic/main.py

```diff
@@ -233,11 +233,7 @@ def __init__(self, **data: Any) -> None:
             self.__values__: Dict[str, Any] = {}
             self.__fields_set__: 'SetStr' = set()
         object.__setattr__(self, '__values__', self._process_values(data))
-        if self.__config__.extra is Extra.allow:
-            fields_set = set(data.keys())
-        else:
-            fields_set = data.keys() & self.__values__.keys()  # type: ignore
-        object.__setattr__(self, '__fields_set__', fields_set)
+        object.__setattr__(self, '__fields_set__', self._process_fields(data))
```

Comment:
humm, I'm not sure this is the easiest or most efficient way of doing things. It might be easier to build `__fields_set__` in `validate_model`.

I think you can change like ~610 to

```py
else:
    names_used.add(field.name if using_name else field.alias)
```
Then add `names_used` along with the other results.

Reply:
`validate_model` currently doesn't mutate the model object at all, so if someone is using it then it may break unexpectedly. There's also no way to return the names_used. If you think it's call `object.__setattr__(model, '__fields_set__', names_used)` from `validate_model` then I can change it to do that.

Reply:
I would change the signature of `validate_model` to return a tuple including `fields_set`.

No one should be using `validate_model` directly so this should be fine.

Reply:
It was a mistake to have to different return types for `validate_model`, best to change it to always return `Tuple['DictStrAny', Set[str], Optional[ValidationError]]`

Reply:
If that's too much work for this PR, I could implement a PR to make the return type always `Tuple['DictStrAny', Optional[ValidationError]]`, then you could rebase off that?

Reply:
Managed to do it. Might need a bit more testing but it should work.

Reply:
Also if no one should be using `validate_model` model it should probably be `_validate_model` or at least not exported in `__init__`.